src/main/java/org/jenkinsci/plugins/workflow/support/pickles/serialization/RiverWriter.java

```diff
@@ -169,26 +164,72 @@ public ObjectOutput getObjectOutput() {
 
     public void close() throws IOException {
         marshaller.finish();
-        int ephemeralsOffset = dout.size();
+        int ephemeralsOffset = (int)channel.position();
 
         // write the ephemerals stream
         pickling = false;
-        marshaller.start(Marshalling.createByteOutput(dout));
+        marshaller.start(new FileChannelOutput(channel));
         marshaller.writeObject(pickles);
         marshaller.finish();
-        dout.close();
 
         // back fill the offset to the ephemerals stream
-        RandomAccessFile raf = new RandomAccessFile(file, "rw");
         try {
-            raf.seek(ephemeralsBackptr);
-            raf.writeInt(ephemeralsOffset);
+            channel.position(EPHEMERALS_BACKPTR);
+            ByteBuffer ephemeralsPtrBuffer = ByteBuffer.allocate(4).putInt(ephemeralsOffset);
+            ephemeralsPtrBuffer.flip();
+            channel.write(ephemeralsPtrBuffer);
+            channel.force(true);
         } finally {
-            raf.close();
+            channel.close();
         }
         LOGGER.fine(() -> "Closed " + file + "; pickle offset @" + ephemeralsOffset);
     }
 
+    private static class FileChannelOutput implements ByteOutput {
+        private final FileChannel channel;
+        /** Used to reduce allocation for single-byte writes. */
+        private final byte[] singleton = new byte[1];
+
+        private FileChannelOutput(FileChannel channel) {
+            this.channel = channel;
+        }
+
+        @Override
+        public void write(int b) throws IOException {
+            singleton[0] = (byte)b; // Downcasting as per interface Javadoc.
+            channel.write(ByteBuffer.wrap(singleton));
+        }
+
+        @Override
+        public void write(byte[] bytes) throws IOException {
+            channel.write(ByteBuffer.wrap(bytes));
+        }
+
+        @Override
+        public void write(byte[] bytes, int off, int len) throws IOException {
+            channel.write(ByteBuffer.wrap(bytes, off, len));
+        }
+
+        @Override
+        public void close() throws IOException {
+            // We only close the channel once all writes are complete.
+        }
+
+        @Override
+        public void flush() throws IOException {
+            // We invoke FileChannel.force manually before closing the channel.
+        }
+    }
+
     /*constant*/ static final long HEADER = 7330745437582215633L;
-    /*constant*/ static final int VERSION = 1;
+    /*constant*/ static final short VERSION = 1;
+    private static final int EPHEMERALS_BACKPTR = 10;
```

Comment:
I think this could have been a constant previously as well since both `DataOutputStream` and `ByteBuffer` are explicit about how many bytes are written for each piece of data.